database/src/main/java/database/service/MySQLConnectionPool.java

```diff
@@ -0,0 +1,109 @@
+package database.service;
+
+import com.mysql.cj.jdbc.MysqlDataSource;
+import database.exception.NoFreeDatabaseConnectionException;
+import database.exception.UnableCreateConnectionException;
+import database.helper.Settings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+public class MySQLConnectionPool {
+    private static final Logger LOG = LoggerFactory.getLogger(MySQLConnectionPool.class);
+    private final MysqlDataSource dataSource;
+    private final Deque<Connection> connectionPool;
+    private final int initialPoolSize;
+    private final int maxPoolSize;
+    private int currentConnections;
+    static final String UNABLE_CREATE_CONNECTION = "Unable to create new database connection.";
+    static final String NO_FREE_DATABASE_CONNECTION = "All connections are in use. Please try again later.";
+
+    public MySQLConnectionPool(Settings settings) {
+        dataSource = createDataSource(settings);
+        this.initialPoolSize = settings.getInitialPoolSize();
+        this.maxPoolSize = settings.getMaxPoolSize();
+        this.connectionPool = new ArrayDeque<>();
```

Comment:
Please remove `this` since we don't use constructor parameters with the same namings.